pkg/kubelet/certificate/certificate_manager.go

```diff
@@ -0,0 +1,371 @@
+	}
+
+	m := manager{
+		certSigningRequestClient: certSigningRequestClient,
+		nodeName:                 nodeName,
+		certDirectory:            filepath.Dir(selectedCertFile),
+		keyDirectory:             filepath.Dir(selectedKeyFile),
+		certRotationEnabled:      certRotationEnabled,
+		cert:                     &cert,
+	}
+
+	if m.certRotationEnabled {
+		if err := m.cleanupPairs(); err == nil {
+			return &m, nil
+		} else {
+			return nil, err
+		}
+	}
+	return &m, nil
+}
+
+// GetCertificate returns the certificate that should be used TLS connections.
+// The value returned by this function will change over time as the certificate
+// is rotated. If a reference to this method is passed directly into the TLS
+// options for a connection, certificate rotation will be handled correctly by
+// the underlying go libraries.
+//
+//    tlsOptions := &server.TLSOptions{
+//        ...
+//        GetCertificate: certificateManager.GetCertificate
+//        ...
+//    }
+//
+func (m *manager) GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	m.certAccessLock.Lock()
+	defer m.certAccessLock.Unlock()
+	return m.cert, nil
+}
+
+// Start will start the background work of rotating the certificates.
+func (m *manager) Start() {
+	if !m.certRotationEnabled {
+		glog.Infof("Certificate rotation is not enabled.")
+		return
+	}
+
+	// Certificate rotation depends on access to the API server certificate
+	// signing API, so don't start the certificate manager if we don't have a
+	// client. This will happen on the master, where the kubelet is responsible
+	// for bootstrapping the pods of the master components.
+	if m.certSigningRequestClient == nil {
+		glog.Infof("Kubernetes client is nil, not starting certificate manager.")
+		return
+	}
+
+	glog.Info("Starting to rotate expiring certificates.")
+	go wait.Forever(func() {
+		for range time.Tick(syncPeriod) {
+			err := m.rotateCerts()
+			if err != nil {
+				glog.Errorf("Could not rotate certificates: %v", err)
+			}
+		}
+	}, 0)
+}
+
+// cleanupPairs is responsible for recovering from any partial certificate
+// rotations rolling back to the most recent matching pair, and deleting any
+// keys that don't have a matching certificate.
+func (m *manager) cleanupPairs() error {
+	if err := m.removeUnmatchedPairs(); err != nil {
+		return err
+	}
+
+	keyPath := filepath.Join(m.keyDirectory, pairNamePrefix+keyExtension)
+	certPath := filepath.Join(m.certDirectory, pairNamePrefix+certExtension)
+	if fileExists(keyPath) && fileExists(certPath) {
+		// Make sure the selected key/cert pair are symlinks. This facilitates
+		// key rotation.
+		if err := fixSymlink(keyPath); err != nil {
+			return err
+		}
+		if err := fixSymlink(certPath); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// removeUnmatchedPairs will remove key files that don't have a matching cert
+// file, and cert files that don't have a matching key file.
+func (m *manager) removeUnmatchedPairs() error {
```

Comment:
> Keeping a single key doesn't address security concerns with regards to a compromised key. Being able to change the key is pretty much the point.

fair

> I can put them in a single file. Do you have a preferred format?

A single file has the upside of a single symlink, no possibility of unmatched files, one file write success and the other failing, etc.

A minor downside is that the single file has to be as protected as the most secret thing in it, which is the key.

I'm pretty sure putting both PEM blocks in a single file works today (can point cert-file and key-file at the same file, and it pulls out the right PEM blocks for each).

> As for auto deleting key files, there is more of that coming. In order to prevent the accumulation of certs and keys over time, old ones will be deleted.

As long as auto-renewed/managed files are separated from user-specified ones, that seems fine.